backends/ubpf/ubpfProgram.cpp

```diff
@@ -0,0 +1,303 @@
+                "static uint64_t (*ubpf_time_get_ns)() = (void *)5;\n"
+                "static uint32_t (*ubpf_hash)(const void *, uint64_t) = (void *)6;\n"
+                "static void (*ubpf_printf)(const char *fmt, ...) = (void *)7;\n"
+                "static void *(*ubpf_packet_data)(const void *) = (void *)9;\n"
+                "static void *(*ubpf_adjust_head)(const void *, uint64_t) = (void *)8;\n"
+                "\n");
+        builder->newline();
+        builder->appendLine(
+                "#define write_partial(a, w, s, v) do { *((uint8_t*)a) = ((*((uint8_t*)a)) "
+                "& ~(BPF_MASK(uint8_t, w) << s)) | (v << s) ; } while (0)");
+        builder->appendLine("#define write_byte(base, offset, v) do { "
+                            "*(uint8_t*)((base) + (offset)) = (v); "
+                            "} while (0)");
+        builder->newline();
+        builder->append("static uint32_t\n"
+                        "bpf_htonl(uint32_t val) {\n"
+                        "    return htonl(val);\n"
+                        "}");
+        builder->newline();
+        builder->append("static uint16_t\n"
+                        "bpf_htons(uint16_t val) {\n"
+                        "    return htons(val);\n"
+                        "}");
+        builder->newline();
+        builder->append("static uint64_t\n"
+                        "bpf_htonll(uint64_t val) {\n"
+                        "    return htonll(val);\n"
+                        "}\n");
+        builder->newline();
+    }
+
+    void UBPFProgram::emitH(EBPF::CodeBuilder *builder, cstring) {
+        emitGeneratedComment(builder);
+        builder->appendLine("#ifndef _P4_GEN_HEADER_");
+        builder->appendLine("#define _P4_GEN_HEADER_");
+        builder->target->emitIncludes(builder);
+        builder->newline();
+        emitTypes(builder);
+        builder->newline();
+        emitTableDefinition(builder);
+        builder->newline();
+        control->emitTableTypes(builder);
+        builder->newline();
+        control->emitTableInstances(builder);
+        builder->appendLine("#endif");
+    }
+
+    void UBPFProgram::emitPreamble(EBPF::CodeBuilder* builder) {
+        builder->emitIndent();
+        builder->appendLine("#define BPF_MASK(t, w) ((((t)(1)) << (w)) - (t)1)");
+        builder->appendLine("#define BYTES(w) ((w) / 8)");
+        builder->newline();
+        builder->appendLine("void* memcpy(void* dest, const void* src, size_t num);");
+        builder->newline();
+    }
+
+    void UBPFProgram::emitTypes(EBPF::CodeBuilder *builder) {
+        for (auto d : program->objects) {
+            if (d->is<IR::Type>() && !d->is<IR::IContainer>() &&
+                !d->is<IR::Type_Extern>() && !d->is<IR::Type_Parser>() &&
+                !d->is<IR::Type_Control>() && !d->is<IR::Type_Typedef>() &&
+                !d->is<IR::Type_Error>()) {
+                if (d->toString().startsWith("struct tuple")) {
```

Comment:
This is weird; what is this supposed to do?

Reply:
Removed. It was useless.